generator/src/generator/namespace.rs

```diff
@@ -5139,19 +5222,49 @@ fn gather_deducible_fields(fields: &[xcbdefs::FieldDef]) -> FxHashMap<String, De
                     if let xcbdefs::Expression::FieldRef(ref field_ref_expr) = switch_field.expr {
                         Some((
                             field_ref_expr.field_name.clone(),
-                            DeducibleField::CaseSwitchExpr(switch_field.name.clone()),
+                            DeducibleField::CaseSwitchExpr(
+                                switch_field.name.clone(),
+                                DeducibleFieldOp::None,
+                            ),
                         ))
                     } else {
-                        None
+                        unreachable!("Can't figure out deducible field of {:#?}", switch_field);
                     }
                 } else if switch_field.kind == xcbdefs::SwitchKind::BitCase {
                     if let xcbdefs::Expression::FieldRef(ref field_ref_expr) = switch_field.expr {
                         Some((
                             field_ref_expr.field_name.clone(),
-                            DeducibleField::BitCaseSwitchExpr(switch_field.name.clone()),
+                            DeducibleField::BitCaseSwitchExpr(
+                                switch_field.name.clone(),
+                                DeducibleFieldOp::None,
+                            ),
                         ))
+                    } else if let xcbdefs::Expression::BinaryOp(ref binary_op_expr) =
+                        switch_field.expr
+                    {
+                        if let xcbdefs::Expression::FieldRef(ref field_ref_expr) =
+                            *binary_op_expr.lhs
+                        {
+                            match binary_op_expr.operator {
+                                xcbdefs::BinaryOperator::And => {
+                                    // An AND X is undone by an OR NOT X.
```

Comment:
TL;DR: I am trying to understand the semantics of `XKB:SelectEvents`.

No, it isn't. `AND FALSE` and `OR NOT FALSE` are not inverse operations. `OR NOT FALSE` always returns in `TRUE`, independent of the input.

Anyway, you make me actually look at the XKB spec. :-)
https://www.x.org/releases/X11R7.7/doc/kbproto/xkbproto.html#Selecting_Events

> The affectWhich , clear , and selectAll fields specify changes to any other event details masks.

Okay, so these fields "go together". We already knew that.

> If any event types are set in both clear and selectAll , a Match error results;

This means that `clear & selectAll == 0` must hold. Good to know.

> if any event types are specified in either clear or selectAll but not in affectWhich , a Match error results.

This means that `clear | selectAll` must only contain bits that are also set in `affectWhich`, so `(clear | selectAll) & ~affectWhich == 0` must hold. This might be the property we actually need...

> Otherwise, the detail masks for any event types specified in the affectWhich field of this request are changed as follows:
> If the event type is also set in clear , the detail mask for the corresponding event is set to 0 or False , as appropriate.
> If the event type is also set in selectAll , the detail mask for the corresponding event is set to include all legal detail values for that type.
> If the event type is not set in either clear or selectAll , the corresponding element of details lists a set of explicit changes to the details mask for the event, as described below.

This third case is what the XML expresses, I think. `affectWhich & (~clear & ~selectAll)` describes which fields are sent.

In the generated code, we know the value of `affectWhich & (~clear & ~selectAll)` because that is what the user explicitly gave us. It makes sense that we want to derive `affectWhich` since this is "the most important" field.

Based on the `(clear | selectAll) & ~affectWhich == 0` from above, it does make sense to do `affectWhich = clear | selectAll | bits_present_in_the_aux_struct`, which is what your code generates. That's good. :-)

However, if some bit is set in the aux structure and also in `clear` or `selectAll`... well, there is nothing that we can do. This is a request that cannot be sent correctly. So, somewhere we still need an assertion.
So... I guess the assertion that is removed below needs to be brought back.

Reply:
> No, it isn't. `AND FALSE` and `OR NOT FALSE` are not inverse operations. `OR NOT FALSE` always returns in `TRUE`, independent of the input.

Yes you're right of course.